lls/cache.c

```diff
@@ -0,0 +1,463 @@
+		else
+			cache->data[ret]->refcnt--;
+	}
+}
+
+unsigned int
+lls_cache_writer(struct lls_cache *cache)
+{
+	struct lls_cache_req *reqs[LLS_CACHE_BURST_SIZE];
+	unsigned int count = mb_dequeue_burst(&cache->mb,
+		(void **)reqs, LLS_CACHE_BURST_SIZE);
+	unsigned int i;
+
+	if (likely(count == 0))
+		return 0;
+
+	for (i = 0; i < count; i++) {
+		switch (reqs[i]->op) {
+		case LLS_CACHE_ADD:
+			__lls_cache_add(cache, reqs[i]->ip_key,
+				reqs[i]->entry);
+			break;
+		case LLS_CACHE_INC:
+			/* FALLTHROUGH */
+		case LLS_CACHE_DEC:
+			__lls_cache_refcnt(cache, reqs[i]->ip_key,
+				reqs[i]->entry, reqs[i]->op);
+			break;
+		default:
+			RTE_LOG(ERR, GATEKEEPER,
+				"lls: unrecognized %s cache opcode (%d)\n",
+				cache->name, reqs[i]->op);
+			break;
+		}
+		mb_free_entry(&cache->mb, reqs[i]);
+	}
+
+	if (DEBUG_LLS_CACHE)
+		lls_cache_dump(cache);
+
+	return count;
+}
+
+int
+lls_cache_get(struct lls_cache *cache, const void *ip_key,
+	struct ether_addr *eth_addr)
+{
+	int ret;
+
+	if (cache == NULL) {
+		RTE_LOG(ERR, GATEKEEPER, "lls: %s: %s cache not instantiated\n",
+			__func__, cache->name);
+		return -EINVAL;
+	}
+
+	ret = rte_hash_lookup(cache->hash, ip_key);
+	if (ret == -ENOENT) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"lls: %s: cannot find entry in %s cache\n",
+			__func__, cache->name);
+	} else if (ret == -EINVAL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"lls: %s: invalid params to lookup entry in %s cache\n",
+			__func__, cache->name);
+	} else {
+		struct lls_cache_entry *lls_entry = cache->data[ret];
+		/* A block has expressed interest, but no resolution exists. */
+		if (!lls_entry->valid)
+			return -ENOENT;
+		ether_addr_copy(&lls_entry->ha, eth_addr);
+		ret = 0;
+	}
+	return ret;
+}
+
+int
+lls_cache_refcnt(struct lls_cache *cache, const void *ip_key, uint8_t port_id,
+	int inc_refcnt)
+{
+	struct lls_cache_entry *lls_entry;
+	struct lls_cache_req *req;
+	int ret;
+
+	if (cache == NULL) {
+		RTE_LOG(ERR, GATEKEEPER, "lls: %s: %s cache not instantiated\n",
+			__func__, cache->name);
+		return -EINVAL;
+	}
+
+	req = mb_alloc_entry(&cache->mb);
+	if (req == NULL) {
+		return -1;
+	}
+
+	/* Allocate entry in case this is a new entry. */
+	ret = rte_mempool_get(cache->entry_pool, (void **)&lls_entry);
+	if (ret == -ENOENT) {
+		mb_free_entry(&cache->mb, req);
+		RTE_LOG(ERR, MEMPOOL,
+			"No memory for new entry, %s entry not added\n",
+			cache->name);
+		return ret;
+	}
+
+	/*
+	 * If no existing entry is found, this one will be used as
+	 * a placeholder that's waiting for resolution. Otherwise,
+	 * this will eventually be discarded.
+	 */
+	lls_entry->ts = time(NULL);
+	lls_entry->port_id = port_id;
+	lls_entry->stale = false;
+	lls_entry->valid = false;
+	lls_entry->refcnt = 1;
+
+	req->entry = lls_entry;
+	req->op = inc_refcnt ? LLS_CACHE_INC : LLS_CACHE_DEC;
+	memcpy(req->ip_key, ip_key, cache->key_len);
+	return mb_send_entry(&cache->mb, req);
+}
+
+int
+lls_cache_add(struct lls_cache *cache, const void *ip_key,
```

Comment:
Given that this function and the above ``lls_cache_refcnt()`` have similar code, can we combine these two functions into one?